wordpress-mobile/WordPress-iOS
# Stats: Fix 'Can't add self as subview' crash in  Posting Activity

Related to https://github.com/wordpress-mobile/WordPress-iOS/issues/20849#issuecomment-1639995735

The majority of new `'Can't add self as subview' crash` [crashes happen in Stats. ](https://github.com/wordpress-mobile/WordPress-iOS/issues/20849#issuecomment-1639995735).

Addressing one of the issues where the user manages to tap "View more" multiple times on `Posting Activity`, resulting in a crash when tapping back multiple times. 

I didn't manage to reproduce the crash myself but I noticed an inefficiency when calling `StatsInsightsStore.getYearlyPostingActivityFrom` method.

`getYearlyPostingActivityFrom` iterates through each day of the year and each existing `PostingStreakEvent` on each iteration, making algorithmic complexity of O(n^2) where n is the number of days in a year.

 I realize that it may not be the most impactful of places to improve Stats so I will look further.

### Solution

Memoize PostingStreakEvent by putting it into the dictionary by date. It reduces the complexity to O(2n) = O(n)
In-app tests have shown the method call being ~14 times faster.

Additionally, I moved the loading of `getYearlyPostingActivityFrom` from the presentation method into `PostingActivityViewModel` and `PostingActivityViewController`, to make sure the data is only loaded when required, not stalling the presentation. 

### Additional improvements

To remove any delays, we could load `getYearlyPostingActivityFrom` in the background asynchronously, presenting a short loading animation in `PostingActivityViewController`. I experimented with it, however with the performance improvements in `getYearlyPostingActivityFrom` I didn't manage to see any benefits. Would you see value in it?

Maybe it's one of the things that could be done with all the Stats: moving any data manipulation into the background and only then making them more efficient one by one, prioritizing snappy navigation + loading animations rather than freezes. 🤔

## To test:

1. Open Jetpack & Log into account
2. Open Stats
3. Add Posting Activity card
4. Confirm that the data appears as before
5. Tap view more
6. Confirm that view opens and data appears as before

## Regression Notes
1. Potential unintended areas of impact

Posting activity data not presented correctly

2. What I did to test those areas of impact (or what existing automated tests I relied on)

Manual tests

3. What automated tests I added (or what prevented me from doing so)

None

PR submission checklist:
